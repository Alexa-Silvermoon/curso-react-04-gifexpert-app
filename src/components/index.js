
export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';
// archivo de barril, tambien conocido como archivo index