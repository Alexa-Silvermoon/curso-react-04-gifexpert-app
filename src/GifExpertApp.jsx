
import { useState } from 'react'
import { AddCategory, GifGrid } from './components/';

// rafc
export const GifExpertApp = () => {

    // arreglo de categorias:
    const [categories, setCategories] = useState(['pc']);
    // NOTA: los hooks no deben ir dentro de condicionales
    // console.log( categories );

    // TODO: boton debe agregar valorant al arreglo de categorias
    // pista: <button onClick={ ( event ) => handleAdd( event, 'hola chaval' ) }> +1 </button>
    const onAddCategory = ( onNewCategory) => {

        // console.log( onNewCategory );

        // si la categoria existe, no la vuelva a incluir en el arreglo
        if ( categories.includes( onNewCategory ) ) return;

        // setCategories( categories.concat('valorant') ); //funciona pero da problemas con lakey
        setCategories([  onNewCategory, ...categories ]); // funciona pero tambien da problemas con la key
        // setCategories( cat => [ ...cat,'valorant']); // funciona pero tambien da problemas con la key


    }

  return (

        <>
            {/* titulo */}
            <h1>GifExpertApp - By Alexander Martinez Millan</h1>

            {/* input */}
            {/* {<AddCategory setCategories={ setCategories }></AddCategory>} */}
            <AddCategory onNewCategory={ ( value ) => onAddCategory( value )} ></AddCategory>

            {/* listado de gif */}
            {/* <button onclick={ onAddCategory }>Agregar</button> */}

            {/* <ol> */}
                {/* { categories.map( category => {

                        return <li key={ category } > { category } </li>
                    })
                } */}

                { categories.map( ( category ) => (

                        /* {/* <div key={ category }>
                            <h3> { category }</h3>
                            <li> { category }</li>
                        </div> */

                        <GifGrid key={ category } category={ category }></GifGrid>
                    ))
                }

            {/* </ol> */}
                {/* gif item */}
    
        </>
    
    );
}

// api key giphy: 14OSyua313kBBUfJYUypqcp7V8PPwZ8x