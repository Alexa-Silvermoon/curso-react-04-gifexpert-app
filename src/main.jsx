import React from 'react'
import ReactDOM from 'react-dom/client'
import { GifExpertApp } from './GifExpertApp'
import './styles.css'; // estilo global

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>

    <GifExpertApp></GifExpertApp>

  </React.StrictMode>
);

// styles: https://gist.github.com/Klerith/e4fca6ac127ccb9abd1e8ad77fcd52f1

/* 

RUTA DE FUNCIONAMIENTO:
main.jsx > GifExpertApp.jsx > GifGrid.jsx > useFetch.js > getGifs.js

luego:

GifGrid.jsx > GifItem.jsx

 */