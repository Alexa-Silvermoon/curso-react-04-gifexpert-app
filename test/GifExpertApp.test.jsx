import { fireEvent, render, screen } from "@testing-library/react";
import { GifExpertApp } from "../src/GifExpertApp";

// TODO: Terminar pruebas en GifExpertApp.jsx

describe('pruebas en GifExpertApp.jsx', () => {

    test('debe agregar una nueva categoria y hacer snapshoot', () => {

        const inputValue = 'chavo';

        const { container } = render( <GifExpertApp></GifExpertApp>);

        expect( container ).toMatchSnapshot();

        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');

        fireEvent.input( input, { target: { value: inputValue } } ); // chavo
        fireEvent.submit( form );

        fireEvent.input( input, { target: { value: inputValue + '2' } } ); // chavo
        fireEvent.submit( form );

        screen.debug();

        expect(screen.getAllByRole('heading', { level: 3 }).length).toBe(3); // heading es etiquetas H
        // NOTA: toBe(3) ya que esta contando PC en const [categories, setCategories] = useState(['pc']);

    });

    test('NO debe agregar una categoria repetida', () => {

        const inputValue2 = 'ramon';

        const { container } = render( <GifExpertApp></GifExpertApp>);

        expect( container ).toMatchSnapshot();

        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');

        fireEvent.input( input, { target: { value: inputValue2 } } ); // ramon
        fireEvent.submit( form );

        fireEvent.input( input, { target: { value: inputValue2 } } ); // ramon
        fireEvent.submit( form );

        expect(screen.getAllByRole('heading', { level: 3 }).length).toBe(2); // heading es etiquetas H
        // NOTA: toBe(2) ya que esta contando PC en const [categories, setCategories] = useState(['pc']);

    });
});

// https://www.udemy.com/course/react-cero-experto/learn/lecture/32021650#questions