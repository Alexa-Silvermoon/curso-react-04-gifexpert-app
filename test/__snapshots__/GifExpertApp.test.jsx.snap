// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en GifExpertApp.jsx NO debe agregar una categoria repetida 1`] = `
<div>
  <h1>
    GifExpertApp - By Alexander Martinez Millan
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar gifs"
      type="text"
      value=""
    />
  </form>
  <h3>
     
    pc
     
  </h3>
  <h2>
    Cargando...
  </h2>
  <div
    class="card-grid"
  />
</div>
`;

exports[`pruebas en GifExpertApp.jsx debe agregar una nueva categoria y hacer snapshoot 1`] = `
<div>
  <h1>
    GifExpertApp - By Alexander Martinez Millan
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar gifs"
      type="text"
      value=""
    />
  </form>
  <h3>
     
    pc
     
  </h3>
  <h2>
    Cargando...
  </h2>
  <div
    class="card-grid"
  />
</div>
`;
